serviceAccount/ServiceAccount.psm1

```diff
@@ -0,0 +1,1211 @@
+            Return $false 
+        } 
+    } 
+     
+    # Clean up the CreateDB.sql file 
+    If (Test-Path "$env:Temp\ADFSSQLScripts\CreateDB.sql") 
+    { 
+        Remove-Item "$env:Temp\ADFSSQLScripts\CreateDB.sql" 
+    } 
+     
+    Return $true 
+     
+} 
+     
+     
+# Executes the SQL scripts generated by GenerateSQLScripts 
+Function ExecuteSQLScripts 
+{ 
+    Start sqlcmd.exe -ArgumentList "-S $SQLHost -i $env:Temp\ADFSSQLScripts\SetPermissions.sql -o $env:Temp\ADFSSQLScripts\SetPermissions.log" -Wait -WindowStyle Hidden | Out-File $LogPath -Append 
+      
+    If (!$?) 
+    { 
+        Write-Host "`tFailed to execute SetPermissions.sql. Exiting" -ForegroundColor "red" 
+        ($ElapsedTime.Elapsed.ToString())+" [ERROR]     Failed to execute SetPermissions.sql" | Out-File $LogPath -Append 
+        Return $false 
+    } 
+      
+     # Execute UpdateServiceSettings.sql, but not for secondary WID. Secondary SQL never gets to this function. 
+      
+    If (!(($Role -eq "SecondaryComputer") -and ($DBMode -eq "WID"))) 
+    { 
+        Start sqlcmd.exe -ArgumentList "-S $SQLHost -i $env:Temp\ADFSSQLScripts\UpdateServiceSettings.sql -o $env:Temp\ADFSSQLScripts\UpdateServiceSettings.log" -Wait -WindowStyle Hidden | Out-File $LogPath -Append 
+      
+        If (!$?) 
+        { 
+            Write-Host "`tFailed to execute UpdateServiceSettings.sql. Exiting...." -ForegroundColor "red" 
+            ($ElapsedTime.Elapsed.ToString())+" [ERROR]     Failed to execute UpdateServiceSettings.sql" | Out-File $LogPath -Append 
+            Return $false 
+        } 
+    } 
+    Return $true 
+} 
+
+
+
+#Define functions to export
+
+
+function Set-AdfsServiceAccountRule
+{
+    param(
+        [parameter(ValueFromPipeline=$True)]
+        [string[]]$SecondaryServers,
+
+        [parameter(Mandatory=$true)]
+        [string]$ServiceAccount
+    )
+
+    #Validate provided account exists
+    $Account = $ServiceAccount.Split('\') #Input should have been given in the format domain\user
+    $User = $Account[1]
+    $Lookup = Get-ADUser -Filter {Name -eq $User} 
+    if($Lookup -eq $null)
+    {
+        Write-Error "The specified account does not exist"
+    }
+
+
+    #Create rule with new service account
+    $SID = ConvertTo-Sid($ServiceAccount)
+    $ServiceAccountRule = "@RuleName = `"Permit Service Account`"`nexists([Type == `"http://schemas.microsoft.com/ws/2008/06/identity/claims/primarysid`", Value == `"$SID`"])`n=> issue(Type = `"http://schemas.microsoft.com/authorization/claims/permit`", value = `"true`");`n`n"
+    $Properties = Get-AdfsInternalSettings
+    $Properties.PolicyStore.AuthorizationPolicy = $Properties.PolicyStore.AuthorizationPolicy + $ServiceAccountRule
+    $Properties.PolicyStore.AuthorizationPolicyReadOnly = $Properties.PolicyStore.AuthorizationPolicyReadOnly + $ServiceAccountRule
+    Set-AdfsInternalSettings  $Properties | out-null
```

Comment:
After you set this on the db, does it require an ADFS restart before the changes take effect? If so, you might want to prompt for the restart

Reply:
No restart is required since the script writes directly to the db

Reply:
The important part is when does ADFS read it from the DB? Do we read the rules on startup and cache them in memory, or do we go to the database every time we need to check the claim rules? My guess is that we cache the rules, but I'm not sure